openssl-dynamic/src/main/c/ssl.c

```diff
@@ -1607,6 +1607,137 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
     return rv;
 }
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+/*
+ * Backport of SSL_SESSION_get_master_key from 1.1
+ */
+size_t SSL_SESSION_get_master_key(const SSL_SESSION *session,
+                                  unsigned char *out, size_t outlen)
+{
+    if (outlen == 0) {
+        return session->master_key_length;
+    }
+    if (outlen > session->master_key_length) {
+        outlen = session->master_key_length;
+    }
+    memcpy(out, session->master_key, outlen);
+    return outlen;
+}
+
+/*
+ * Backport of SSL_get_server_random from 1.1
+ */
+size_t SSL_get_server_random(const SSL *ssl, unsigned char *out, size_t outlen)
+{
+    if (outlen == 0) {
+        return sizeof(ssl->s3->server_random);
+    }
+    if (outlen > sizeof(ssl->s3->server_random)) {
+        outlen = sizeof(ssl->s3->server_random);
+    }
+    memcpy(out, ssl->s3->server_random, outlen);
+    return outlen;
+}
+
+/*
+ * Backport of SSL_get_client_random from 1.1
+ */
+size_t SSL_get_client_random(const SSL *ssl, unsigned char *out, size_t outlen)
+{
+    if (outlen == 0) {
+        return sizeof(ssl->s3->client_random);
+    }
+    if (outlen > sizeof(ssl->s3->client_random)) {
+        outlen = sizeof(ssl->s3->client_random);
+    }
+    memcpy(out, ssl->s3->client_random, outlen);
+    return outlen;
+}
+#endif
+
+TCN_IMPLEMENT_CALL(jbyteArray, SSL, getClientRandom)(TCN_STDARGS, jlong ssl)
+{
+
+    SSL *ssl_ = J2P(ssl, SSL *);
+    TCN_CHECK_NULL(ssl_, ssl, NULL);
+
+    size_t keyLength = SSL_get_client_random(ssl_, NULL, 0);
+    TCN_ASSERT(keyLength <= 0x7FFFFFFF); /* must fit into 32 bit unsigned jsize */
+
+    unsigned char *key = OPENSSL_malloc(sizeof(unsigned char) * keyLength);
+    if (key == NULL) {
+        tcn_ThrowException(e, "OPENSSL_malloc() returned null");
+        return NULL;
+    }
+
+    size_t bytesMoved = SSL_get_client_random(ssl_, key, keyLength);
+    TCN_ASSERT(bytesMoved == keyLength);
+
+    jbyteArray jKey = (*e)->NewByteArray(e, (jsize) bytesMoved);
+    (*e)->SetByteArrayRegion(e, jKey, 0, bytesMoved, (jbyte*) key);
```

Comment:
I think we need to guard against when `NewByteArray(...)` returns `NULL` (like if there is no memory left).

Reply:
I can add the NULL checks however I stuck to previous convention where the other JNI calls did no such guard.